### ISSUE TITLE
Reflected braille dots

### ISSUE THREAD

**Issue by winter-yuki** [braille-systems/learn-braille#223]
https://github.com/braille-systems/learn-braille/issues/221#issuecomment-664232669

**Comment by winter-yuki** [comment]
It looks like some braille dots design changes are needed to represent input mode.

**Comment by zuevval** [comment]
Now it is more or less clear what Victoria Nikolaevna (and, presumably, other visually impaired) would like. See [wiki article, point 8](https://github.com/braille-systems/learn-braille/wiki/2020-august-roadmap).

> It looks like some braille dots design changes are needed to represent input mode.

Quite agree. In the text (see link) in the same point 8 I proposed demarcating dot 1. Tomorrow I plan to try to implement it.

**Comment by zuevval** [comment]
which one is better?

<img src="https://user-images.githubusercontent.com/23435506/90366962-bc664780-e070-11ea-8d80-fc63481efecb.png" height=400>  <img src="https://user-images.githubusercontent.com/23435506/90366966-bcfede00-e070-11ea-9195-dee18a18cf54.png" height=400>

**Comment by zuevval** [comment]
This will do, er?

<img src="https://user-images.githubusercontent.com/23435506/90369799-0b15e080-e075-11ea-963d-bc3900a25811.png" height="280">  <img src="https://user-images.githubusercontent.com/23435506/90369804-0bae7700-e075-11ea-96e6-8f5ab731f237.png" height="280">

**Comment by winter-yuki** [comment]
> which one is better?
> 
> <img alt="" height="400" src="https://user-images.githubusercontent.com/23435506/90366962-bc664780-e070-11ea-8d80-fc63481efecb.png"> <img alt="" height="400" src="https://user-images.githubusercontent.com/23435506/90366966-bcfede00-e070-11ea-9195-dee18a18cf54.png">

Right one looks better to me

**Comment by winter-yuki** [comment]
> This will do, er?
> 
> <img alt="" height="400" src="https://user-images.githubusercontent.com/23435506/90369799-0b15e080-e075-11ea-963d-bc3900a25811.png"> <img alt="" height="400" src="https://user-images.githubusercontent.com/23435506/90369804-0bae7700-e075-11ea-96e6-8f5ab731f237.png">

I do not see any necessity to add this button everywhere directly on the screen. I'd like to add an item in settings to reflect dots in the practice and in the browser.

**Comment by zuevval** [comment]
well, why, in books for sighted they even go in pair. I will think about it, but maybe at least in browser we'd better place a button

**Comment by zuevval** [comment]
I’ve asked a question about practice in WhatsApp. @winter-yuki feel free to add something if you think that my question is ill-posed

**Comment by zuevval** [comment]
So we received two answers from which I conclude that the button would be, most likely, OK. <br>
Now I suppose we should add flip via button and roll out beta release, then ask Viktoria Nikolaevna to test and tell whether it's OK.

How to do it technically? I suppose, next steps are:
1. in all input steps in lessons: flip columns
1. add a new type of show steps (or an optional argument for constructor) - with flipped columns
1. in browser, flip dots on button click (default: reading mode, state is not saved)
1. in practice, flip dots on button click (default: writing mode, state is not saved)
1. change help messages (and maybe make different titles for different show steps)

@winter-yuki I would rather ask you to do it (except for the last step). I've tried to do something, e. g. change `brailleDots` background on button click, but realized I know too few of Android and how it all should be implemented. Maybe it's time to create a custom view for `brailleDots` with methods like `flipToFront` and `flipToBack`?

**Comment by winter-yuki** [comment]
Adding this button closer to the user's position is useless in general, type of input won't change for every card or browser's view. So I think that its purpose (being available everywhere) is to remind users about dots mode via talkback.

**Comment by winter-yuki** [comment]
The question in WhatsApp was not totally correct. It is a big difference between placing this button in the context menu above or directly on the screen. Also received answers have no argumentation. Having it on the screen is useless and inconsistent to my opinion. Being placed in the context menu, it will solve the main problem of reminding users about the current input mode.

**Comment by winter-yuki** [comment]
Should buttons in practice and in browser flip the same preferences item or switches should be independent?

**Comment by winter-yuki** [comment]
> So we received two answers from which I conclude that the button would be, most likely, OK.
> 
> Now I suppose we should add flip via button and roll out beta release, then ask Viktoria Nikolaevna to test and tell whether it's OK.
> 
> How to do it technically? I suppose, next steps are:
> 
>     1. in all input steps in lessons: flip columns
> 
>     2. add a new type of show steps (or an optional argument for constructor) - with flipped columns
> 
>     3. in browser, flip dots on button click (default: reading mode, state is not saved)
> 
>     4. in practice, flip dots on button click (default: writing mode, state is not saved)
> 
>     5. change help messages (and maybe make different titles for different show steps)
> 
> 
> @winter-yuki I would rather ask you to do it (except for the last step). I've tried to do something, e. g. change `brailleDots` background on button click, but realized I know too little of Android and how it all should be implemented. Maybe it's time to create a custom view for `brailleDots` with methods like `flipToFront` and `flipToBack`?

1. Do you want to flip dots for each step? Doing it implicitly is not a good idea for current users. Maybe flip dots for new users and provide some kind of switch for others?
2. Maybe add method like `reflected() :: BrailleDots -> BrailleDots`?

**Comment by winter-yuki** [comment]
I did not notice items 3 and 4. I do not see any purposes of flipping dots for the current card or session. If not to get used to the mode then what for is it all?

**Comment by zuevval** [comment]
When a student punches holes (on reverse side) with slate and stylus, he may flip the paper and see the front side.

**Comment by zuevval** [comment]
I'm sorry for mixing two ideas:
1. flipping the paper (to check the result of "writing")
2. practicing Braille via typing symbols "as in reading mode", without "writing" at all

if the goal is (1), it seems to me more logical to place a BIG BUTTON to the right (as now), but prohibit editing while card is flipped to front.

By the way, I'm back to Petersburg; maybe we discuss it by phone?

**Comment by winter-yuki** [comment]
Ok, better by zoom

**Comment by zuevval** [comment]
So we've chatted and consented:
1. **theory:** all input - *as while writing*, `showSymbol` expands to 2 steps: first *as while writing*, second *as while reading*
1. **practice:** all input - *as while writing*, but may be changed to *as while reading* and back via "flip" button
1. **materials browser:** default view - *as while writing*, may be changed to *as while reading* and back via "flip" button
4. there will be a switch in settings which will change default view in practice and materials browser to *as while reading*. **During the call I said that it should not alter the input in theory, but now I thought more and changed my mind: maybe it's sensible to shift the input in theory to *as while reading*?**

ℹ️ flipping card means:
1. columns interchange visually
1. (in practice or input step) columns interchange on input
1. background changes
1. accessibility labels of dots are swapped 

ℹ️  *as while writing* = reversed (123 on the **right**, 456 on the **left**), *as while reading* = as now (123 on the **left**, 456 on the **right**)

**Comment by winter-yuki** [comment]
@zuevval 1. maybe add possibility in settings to disable repeating show steps with reflected dots?

**Comment by zuevval** [comment]
> @zuevval 1. maybe add possibility in settings to disable repeating show steps with reflected dots?

There are not so many show steps in the course. Each symbol is shown only once. And it's important for users to understand and get accustomed to the concept of flipped columns. So I don't think it's a good idea

**Comment by zuevval** [comment]
@winter-yuki maybe somewhere at the very bottom of the list if you think it's necessary

**Comment by winter-yuki** [comment]
> @winter-yuki maybe somewhere at the very bottom of the list if you think it's necessary. But I think it's not necessary.

It is rather simple. If it can be useful, I'll do it

**Comment by winter-yuki** [comment]
@zuevval are there explanations about braille reading and writing mode in the `golubina` course?

**Comment by zuevval** [comment]
> @zuevval are there explanations about braille reading and writing mode in the `golubina` course?

yes, I have added them already in the first lesson. But there is still some work to do with help messages

**Comment by winter-yuki** [comment]
@zuevval should dots traversal order be reflected?

**Comment by zuevval** [comment]
> @zuevval should dots traversal order be reflected?

Yes, because otherwise, those who use TalkBack mostly with swipes will not notice any difference

**Comment by zuevval** [comment]
🎺
